File: local-gateway/static/js/components/workflows.js
/* local-gateway/static/js/components/workflows.js */
import { apiGet, apiPost, toast } from '../api.js';
import { formatTime, formatTimeShort, escapeHtml } from '../utils.js';

document.addEventListener('alpine:init', () => {
  Alpine.data('workflows', () => ({
    loading: true,
    error: null,
    workflows: [],

    // Create form
    showForm: false,
    form: { name: '', description: '', trigger_type: 'manual', trigger_config: '{}', actions: '[]' },

    // Trigger/action type metadata
    triggerTypes: [],
    actionTypes: [],

    // Execution history
    showHistory: false,
    historyWorkflowId: null,
    historyWorkflowName: '',
    historyLoading: false,
    historyItems: [],

    // Confirm delete
    confirmDeleteId: null,

    async init() {
      this.$watch('$store.view.current', (val) => {
        if (val === 'workflows') this.load();
      });
      if (Alpine.store('view').current === 'workflows') await this.load();
    },

    async load() {
      this.loading = true;
      this.error = null;
      try {
        const data = await apiGet('/api/workflows');
        this.workflows = Array.isArray(data) ? data : (data.workflows || data.data || []);
      } catch (e) {
        this.error = e.message || '加载失败';
        toast('加载工作流失败', 'error');
      } finally {
        this.loading = false;
      }
    },

    openNewForm() {
      this.form = { name: '', description: '', trigger_type: 'manual', trigger_config: '{}', actions: '[]' };
      this.showForm = true;
    },

    async createWorkflow() {
      if (!this.form.name) {
        toast('请输入工作流名称', 'error');
        return;
      }
      let triggerConfig, actions;
      try {
        triggerConfig = JSON.parse(this.form.trigger_config);
        actions = JSON.parse(this.form.actions);
      } catch {
        toast('触发配置或动作必须是有效的 JSON', 'error');
        return;
      }
      try {
        const data = await apiPost('/api/workflows', {
          name: this.form.name,
          description: this.form.description,
          trigger_type: this.form.trigger_type,
          trigger_config: triggerConfig,
          actions: actions,
        });
        if (data.status !== 'success' && data.status !== undefined && data.status !== 'ok') {
          throw new Error(data.message || '创建失败');
        }
        toast('工作流已创建', 'success');
        this.showForm = false;
        await this.load();
      } catch (e) {
        toast('创建工作流失败', 'error');
      }
    },

    async toggleWorkflow(id) {
      const wf = this.workflows.find(w => w.id === id || w.workflow_id === id);
      const oldEnabled = wf ? wf.enabled : true;
      if (wf) wf.enabled = !wf.enabled;

      try {
        const data = await apiPost(`/api/workflows/${id}/toggle`, {});
        if (data.status !== 'success' && data.status !== 'ok' && data.status !== undefined) {
          throw new Error(data.message || '切换失败');
        }
        toast(wf?.enabled ? '工作流已启用' : '工作流已禁用', 'success');
      } catch (e) {
        if (wf) wf.enabled = oldEnabled;
        toast('切换失败', 'error');
      }
    },

    async executeWorkflow(id, e) {
      e.stopPropagation();
      const wf = this.workflows.find(w => w.id === id || w.workflow_id === id);
      toast(`正在执行「${escapeHtml(wf?.name || '')}」...`, 'info');
      try {
        const data = await apiPost(`/api/workflows/${id}/execute`, {});
        if (data.status !== 'success' && data.status !== 'ok' && data.status !== undefined) {
          throw new Error(data.message || '执行失败');
        }
        toast('工作流已执行', 'success');
      } catch (e) {
        toast('执行失败', 'error');
      }
    },

    async deleteWorkflow(id) {
      if (!confirm('确定删除此工作流？')) return;
      this.workflows = this.workflows.filter(w => (w.id !== id && w.workflow_id !== id));

      try {
        const resp = await fetch(`/api/workflows/${id}`, { method: 'DELETE' });
        const data = await resp.json();
        if (data.status !== 'success' && data.status !== 'ok' && data.status !== undefined) {
          throw new Error(data.message);
        }
        toast('工作流已删除', 'success');
      } catch (e) {
        toast('删除失败', 'error');
        await this.load();
      }
    },

    async loadHistory(id, name) {
      if (this.showHistory && this.historyWorkflowId === id) {
        this.showHistory = false;
        return;
      }
      this.historyWorkflowId = id;
      this.historyWorkflowName = name;
      this.historyLoading = true;
      try {
        const data = await apiGet(`/api/workflows/${id}/executions`);
        this.historyItems = Array.isArray(data) ? data : (data.executions || data.data || []);
        this.showHistory = true;
      } catch (e) {
        toast('加载执行历史失败', 'error');
        this.historyItems = [];
      } finally {
        this.historyLoading = false;
      }
    },

    closeHistory() {
      this.showHistory = false;
      this.historyWorkflowId = null;
      this.historyWorkflowName = '';
      this.historyItems = [];
    },

    triggerLabel(type) {
      const map = { manual: '手动', schedule: '定时', webhook: 'Webhook', event: '事件' };
      return map[type] || type || '手动';
    },

    getWorkflowId(wf) {
      return wf.id || wf.workflow_id;
    },

    statusLabel(status) {
      const map = { success: '成功', failed: '失败', running: '运行中', pending: '等待中' };
      return map[status] || status || '-';
    },

    statusClass(status) {
      if (status === 'success') return 'badge-completed';
      if (status === 'failed') return 'badge-error';
      return 'badge-pending';
    },

    formatTime: formatTime,
    formatTimeShort: formatTimeShort,
    escapeHtml: escapeHtml,
  }));
});

File: local-gateway/routers/workflows.py
"""
工作流路由 — 自动化工作流管理
"""
from fastapi import APIRouter, Query

from models.schemas import BaseModel, Field
from services.workflow_service import (
    ACTION_TYPES,
    TRIGGER_TYPES,
    workflow_engine,
)

router = APIRouter(prefix="/workflows", tags=["workflows"])


class WorkflowCreateRequest(BaseModel):
    name: str = Field(..., description="工作流名称")
    description: str = Field("", description="工作流描述")
    trigger: dict = Field(..., description="触发器配置")
    actions: list[dict] = Field(..., description="动作列表")
    enabled: bool = Field(True, description="是否启用")


class WorkflowTriggerRequest(BaseModel):
    context: dict = Field({}, description="上下文数据")


@router.get("/")
async def list_workflows():
    """获取所有工作流"""
    return workflow_engine.get()


@router.post("/")
async def create_workflow(request: WorkflowCreateRequest):
    """创建工作流"""
    return workflow_engine.create(
        name=request.name,
        trigger=request.trigger,
        actions=request.actions,
        description=request.description,
        enabled=request.enabled,
    )


@router.get("/{workflow_id}")
async def get_workflow(workflow_id: str):
    """获取工作流详情"""
    return workflow_engine.get(workflow_id)


@router.delete("/{workflow_id}")
async def delete_workflow(workflow_id: str):
    """删除工作流"""
    return workflow_engine.delete(workflow_id)


@router.post("/{workflow_id}/toggle")
async def toggle_workflow(
    workflow_id: str,
    enabled: bool = Query(..., description="启用/禁用"),
):
    """启用/禁用工作流"""
    return workflow_engine.toggle(workflow_id, enabled)


@router.post("/{workflow_id}/execute")
async def execute_workflow(
    workflow_id: str,
    request: WorkflowTriggerRequest,
):
    """手动执行工作流"""
    return await workflow_engine.execute(workflow_id, request.context)


@router.get("/{workflow_id}/executions")
async def get_workflow_executions(
    workflow_id: str,
    limit: int = Query(50, ge=1, le=100),
):
    """获取工作流执行记录"""
    return workflow_engine.get_executions(workflow_id, limit)


@router.get("/types/triggers")
async def list_trigger_types():
    """获取支持的触发器类型"""
    return {
        "status": "success",
        "types": TRIGGER_TYPES,
    }


@router.get("/types/actions")
async def list_action_types():
    """获取支持的动作类型"""
    return {
        "status": "success",
        "types": ACTION_TYPES,
    }


@router.post("/trigger/{trigger_type}")
async def trigger_workflows(
    trigger_type: str,
    context: dict = {},
):
    """
    触发指定类型的工作流

    用于手动触发或通过 Webhook 触发
    """
    results = await workflow_engine.trigger(trigger_type, context)
    return {
        "status": "success",
        "triggered": len(results),
        "results": results,
    }
